rfd/0181-pg-database-access-webui.md

```diff
@@ -0,0 +1,322 @@
+  // sessions.
+  SupportsWebUISession bool `json:"support_webui_session"`
+}
+```
+
+These new fields are used to:
+- Change the "Connect" button behavior to present a connect modal when web UI
+  is supported.
+- Present the available database roles when applicable.
+
+In addition, a new page will be added to handle the database session
+console/terminal (`/web/cluster/:clusterId/console/db/:sid`). This page will be
+very similar to the SSH session console, except for the file transfer/upload
+icons bar, which will be removed.
+
+#### Proxy
+
+A new web handler will be added to handle the interactive session socket
+connection: `/webapi/sites/:site/databases/term`. We cannot provide the session
+parameters since web socket connections are started as GET requests. Instead,
+when the web socket connection is established, the front end will forward a
+message containing the session parameters. This is the same flow used by
+Kubernetes exec sessions.
+
+The following struct represents the session request:
+
+```go
+// DatabaseSessionRequest describes a request to create a web-based terminal
+// database session.
+type DatabaseSessionRequest struct {
+  // DatabaseResourceID is the database resource ID the user will be connected.
+  DatabaseResourceID string `json:"db_resource_id"`
+  // DatabaseName is the database name the session will use.
+  DatabaseName string `json:"db_name"`
+  // DatabaseUser is the database user used on the session.
+  DatabaseUser string `json:"db_user"`
+  // DatabaseRoles are ratabase roles that will be attached to the user when connecting to the database.
+  DatabaseRoles []string `json:"db_roles"`
+}
+```
+
+After the connection is established and the first message is received, the proxy
+will generate a user certificate containing `RouteToDatabase` using one of
+`PerformMFACeremony` or `GenerateUserCerts`.
+
+Note: The MFA flows will follow the existing functions (used by Kubernetes
+execs, desktops, and SSH sessions).
+
+Once the certificates are generated, the proxy will connect to the target
+database server. This flow resembles what the database proxy does: Choose a
+database server among a list of available servers to forward the connection.
+After the connection is established, a new instances of the Teleport's
+PostgreSQL REPL is created, and the session starts.
+
+#### PostgreSQL REPL
+
+The proxy will include a PostgreSQL Read-Eval-Print Loop (REPL) that interacts
+with the databases served by database servers. The REPL will read the user input
+per line and decide whether or not to execute the command (as the REPL will
+support multi-line commands). The rule follows:
+- If the line starts with a backslash command, execute it.
+- Send the line to the target database if it terminates in a semicolon (`;`).
+
+If the line read doesn't meet the above criteria, it is added to an internal
+buffer and is accumulated until it is ready to be executed.
```

Comment:
What happens for the following input:
```sql
SELECT * FROM
\du
pg_roles;
```

First line is added to buffer, but second line is then "executed". As third line is terminated with `;`, what will be sent?

What happens if there are multiple `;` separated commands in a single line?